### PR TITLE
Optionally construct Shelley bootstrap witnesses with derivation path

### LINKED ISSUE #1851: [BUG] - cardano-cli shelley transaction sign: missing derivation path for Byron-random bootstrap witnesses
**Internal/External**
*Internal*   if an IOHK staff member.
~~*External*   otherwise.~~

**Summary**
A clear and specific description of what the bug is.

`makeShelleyBootstrapWitness` from `Cardano.Api.Typed` assume that all Byron addresses have an empty derivation path. That is only true for so-called 'Icarus' addresses that came with Yoroi. Old cardano-sl / Daedalus Byron addresses do include an encrypted derivation path. 

https://github.com/input-output-hk/cardano-node/blob/9c8130eb79332391a8fecc2e4fda84e94239ef8e/cardano-api/src/Cardano/Api/Typed.hs#L1321-L1326

As a consequence, the address root being computed and stored with the witness is wrong and end up "confusing" the ledger unable to identify witnesses correctly. 

**Steps to reproduce**

1. Send some funds to a Byron-random address (Testnet is fine).  Note, you can distinguish Icarus and Random addresses by using the `cardano-address address inspect` command as such:

```console
$ cardano-address address inspect <<< 37btjrVyb4KG1XNuagboo1RD5zpMkHaJ4B9J7fZ2fNPfzxsYy28USBow2DmkTkF8QP8d2Lxb4sdSeWx4D6G4m9CZrkDVYpzqrJSK9Fm1WWtGAz86Ek
{
    "stake_reference": "none",
    "address_style": "Byron",
    "address_root": "f0598e0f99c6b1778b29aee326ae7ace421aec2facac0cf702dfedc3",
    "derivation_path": "581c6e52a48e843a966588b5d6c2d014fda7f02e6260ef83fb06c748db72",
    "network_tag": 1097911063
}
```

And notice the non-null `derivation_path` attribute as well as the Byron `address_style`.

2. Prepare a transaction using `cardano-cli shelley transaction build-raw ...` that is trying to spend a UTxO at that address.

3. Sign the transaction using the private key corresponding to that address `cardano-cli shelley transaction sign ...`

4. Submit the signed transaction `cardano-cli shelley transaction submit ...`

**Expected behavior**

The transaction should be submitted just fine (all other details assumed correct), but it is rejected by the ledger with an error similar to:

```
Shelley command failed: transaction submit  Error: Error while submitting tx: ApplyTxError [LedgerFailure (UtxowFailure (MissingVKeyWitnessesUTXOW (WitHashes (fromList [KeyHash "f0598e0f99c6b1778b29aee326ae7ace421aec2facac0cf702dfedc3"]))))]
```

**System info (please complete the following information):**
- OS: All
- Version: Visible on current master, 1.19.0 and 1.19.1 (and probably prior to this).
- Node version: -

**Screenshots and attachments**
- If applicable, add screenshots, config files and/or logs to help explain the problem.

**Additional context**

See also with @disassembler and Neville for reproduction.

### PR DESCRIPTION
**Pull request by intricate**
Closes #1851

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs [hunk 1]
```diff
@@ -306,16 +307,16 @@ data SomeWitnessSigningKey
                                (Api.SigningKey Api.GenesisDelegateExtendedKey)
   | AGenesisUTxOSigningKey     (Api.SigningKey Api.GenesisUTxOKey)
 
-readSigningKeyFile
-  :: SigningKeyFile
+readWitnessSigningData
+  :: WitnessSigningData
   -> ExceptT (Api.FileError SigningKeyDecodeError) IO SomeWitnessSigningKey
-readSigningKeyFile skFile =
+readWitnessSigningData (WitnessSigningData skFile mbByronAddr) =
```
Comment (intricate) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:313]: I should do better validation in this function. 

If an `Address Byron` is provided, an error should be returned if any other than a `AByronSigningKey` is deserialized from the `SigningKeyFile`.

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 2]
```diff
@@ -1467,6 +1478,21 @@ pFilterByStakeAddress =
       <> Opt.help "Filter by Cardano stake address (Bech32-encoded)."
       )
 
+pByronAddress :: Parser (Address Byron)
+pByronAddress =
+    Opt.option
+      (Opt.eitherReader deserialise)
+        (  Opt.long "address"
+        <> Opt.metavar "STRING"
+        <> Opt.help "Byron address (Base58-encoded)."
```
Comment (intricate) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:1487]: Improve this help text to explain why one might want to specify a Byron address when constructing a bootstrap witness.

**Review by Jimbo4350 — APPROVED** [review]

**Review by Jimbo4350 — APPROVED** [review]

**Review by newhoggy — APPROVED** [review]

**Comment by newhoggy** [comment]
Needs rebase

**Pull request merged**